Capability discovery

j0sh opened the pull request:
**What does this pull request do? Explain your changes. (required)**

Implements the first stages of capability discovery - https://github.com/livepeer/go-livepeer/issues/1519

This does not implement (and will not implement):

* Additional constraints. Only stubs out the structures.
* Capability storage in broadcaster DB
* Capability discovery for standalone T

**Specific updates (required)**
<!--- List out all significant updates your code introduces -->
- net: Regenerate proto file.  c2aed29
- core: Add support for capability enumeration.  2a3f05a
- discovery: Add capability discovery. cba60a5
- core: Add encoding profiles to capabilities. bef2014 Added to demonstrate how to add a new capability. Note that the list of profiles is explicit, if one ever needs to be disabled or superseded.
- core: Add GOP to capabilities. cb4d6ae Added to demonstrate how to add a new capability.
- core: Add mandatory capability support. efdc188 See https://github.com/livepeer/go-livepeer/issues/1519 for a description of orchestrator mandatory capabilities. Currently, nothing is mandatory, but some things might be soon (eg, transcoding protocol authentication - https://github.com/livepeer/go-livepeer/issues/1569 )
-server: Add orchestrator checking of segment capabilities. … e79ec85 . Acts as a validation check prior to segment processing, and needed for capability matching with standalone T (not currently supported).

**How did you test each of these updates (required)**
<!-- A detailed description of how you tested your code changes. Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->

* Unit testing
* Regression testing via https://gist.github.com/j0sh/9d884a19bb92e504092e71fd19d812e5
* Manual testing, including  builds with certain capabilities intentionally disabled to induce failures. (There is currently no way to selectively disable capabilities outside of compile time.)
  * Capability Discovery B,  Old O
  * Old B, Capability Discovery O
  * Capability Discovery B, Capability Discovery O, All Features Enabled
  * Capability Discovery B, Capability Discovery O, mpegts disabled, MP4 transcoding request
  * Capability Discovery B, Capability Discovery O, mpegts disabled, mpegts transcoding request
  * [ ] TODO Test a build with object storage disabled, and external OS in use


**Does this pull request close any open issues?**
<!-- Fixes # -->

Fixes https://github.com/livepeer/go-livepeer/issues/1519

**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] README and other documentation updated
- [ ] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

j0sh committed:
core: Add support for capability enumeration.

j0sh committed:
discovery: Add capability discovery.

j0sh commented on core/capabilities.go:
```diff
@@ -0,0 +1,166 @@
+	Capability_FractionalFramerates
+	Capability_StorageDirect
+	Capability_StorageS3
+	Capability_StorageGCS
```
Figured it'd be good to be as specific as possible about all the features that are being used, including the storage driver, in case we ever need to disable support for one of these someday.

jailuthra reviewed (approved):
LGTM!

kyriediculous commented:
Just lurking here 👀  

Uttermost interested in "Capability discovery for standalone T"

Should we consider gitignoring the generated grpc go stubs because they're part of the makefile anyhow (I notice a conflict) ?

j0sh commented:
> Should we consider gitignoring the generated grpc go stubs

@kyriediculous Do you mean the `net/lp_rpc.pb.go` file ? That would require the protobuf-go compiler be installed - it's a bit easier sometimes without that. Also, different versions of the compiler can cause unwanted changes or even breakage as @jailuthra encountered recently [1]. Checking in the generated code allows us to punt on the issue and ensure everything is synchronized for everyone.

[1] We should definitely still fix these types of issues eventually, maybe at the same time we fix go 1.14 😅  https://github.com/livepeer/go-livepeer/pull/1414

j0sh committed:
net: Regenerate proto file.

j0sh committed:
core: Add encoding profiles to capabilities.

j0sh committed:
core: Add GOP to capabilities.

j0sh committed:
core: Add mandatory capability support.

j0sh committed:
server: Add orchestrator checking of segment capabilities.

Needed for capability matching with standalone T, and acts as a
validation check prior to segment processing.

j0sh commented:
Rebased against master and added a few more things.

Profiles and GOP support (separately) to demonstrate how to add new capabilities.

Mandatory capabilities because this allows for "forward compatibility" - even though we don't have any such mandatory capabilities right now, broadcasters that know to check the field will be able to reject orchestrators that add mandatory capabilities in the future, even if the broadcasters themselves are not aware of the capability. See https://github.com/livepeer/go-livepeer/issues/1519 for more details.

Orchestrator checking of job capabilities - because this is going to be needed at some point, and it's a good sanity check on the orchestrator side.

Still needs a little more manual testing, especially around object storage, but things are ready for review. This PR is as complete as it's going to get, feature-wise.

yondonfu commented on core/capabilities.go:
```diff
@@ -0,0 +1,246 @@
+
+	// generate bitstring
+	capList := []Capability{}
+	for k, _ := range caps {
```
Should be able to shorten this to `for k := range caps`

j0sh replied:
Fixed in 176c77e0d5ec5b7cd30e1cb13746ee16728d9513

yondonfu commented on core/capabilities.go:
```diff
@@ -0,0 +1,246 @@
+	}
+
+	// For now, check this:
+	// ( orch.mandatories  AND orch.bitstring  ) == orch.mandatories &&
```
Is this comment accurate? Based on the checks below I would think that this comment should say:

```
( orch.mandatories AND bcast.bitstring ) == orch.mandatories &&
( bcast.bitstring AND orch.bitstring ) == bcast.bitstring
```

j0sh replied:
Yep, typos. Fixed in 192981f3893798f656f0ea973d4b45c48a2e63a6

yondonfu commented on server/rpc.go:
```diff
@@ -340,12 +342,20 @@ func coreSegMetadata(segData *net.SegData) (*core.SegTranscodingMetadata, error)
+	caps := core.CapabilitiesFromNetCapabilities(segData.Capabilities)
+	if caps == nil {
+		// For older broadcasters. Note if there are any orchestrator
+		// mandatory capabilities, segment verification will fail.
```
Should "segment verification will fail" say something like "seg creds verification will fail" since I believe it is referring to the compatibility check in `verifySegCreds()`?

j0sh replied:
Fixed in 2eca7be58e2990dabf716372c7734bee1e7d992f

yondonfu reviewed (commented):
Looks great just a few minor comments.

yondonfu reviewed (approved):
LGTM :shipit:
